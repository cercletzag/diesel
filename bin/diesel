#!/bin/sh
cd diesel_cli && cargo run -- $@
