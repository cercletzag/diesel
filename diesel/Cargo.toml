[package]
name = "diesel"
version = "0.4.1"
authors = ["Sean Griffin <sean@seantheprogrammer.com>"]
license = "MIT OR Apache-2.0"
description = "A safe, extensible ORM and Query builder"
documentation = "http://sgrif.github.io/diesel/diesel/index.html"
homepage = "https://github.com/sgrif/diesel"
repository = "https://github.com/sgrif/diesel"
keywords = ["orm", "database", "postgres", "postgresql", "sql"]

[dependencies]
libc = "0.2.*"
pq-sys = { version = "^0.2.0", optional = true }
libsqlite3-sys = { version = "^0.4.0", optional = true }
byteorder = "0.3.*"
quickcheck = { git = "https://github.com/BurntSushi/quickcheck.git", optional = true }
chrono = { version = "^0.2.17", optional = true }

[dev-dependencies]
quickcheck = { git = "https://github.com/BurntSushi/quickcheck.git" }
dotenv = "^0.6.0"
tempdir = "^0.3.4"

[features]
default = ["postgres"]
unstable = []
large-tables = []
postgres = ["pq-sys"]
sqlite = ["libsqlite3-sys"]
