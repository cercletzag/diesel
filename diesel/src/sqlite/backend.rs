use backend::*;
use super::connection::SqliteValue;
use super::query_builder::SqliteQueryBuilder;

pub struct Sqlite;

pub enum SqliteType {
    Binary,
    Text,
    Float,
    Double,
    SmallInt,
    Integer,
    Long,
}

impl Backend for Sqlite {
    type QueryBuilder = SqliteQueryBuilder;
    type RawValue = SqliteValue;
}

impl TypeMetadata for Sqlite {
    type TypeMetadata = SqliteType;
}
