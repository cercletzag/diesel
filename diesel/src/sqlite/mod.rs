mod backend;
mod connection;
mod query_builder;
mod types;

pub use self::backend::{Sqlite, SqliteType};
pub use self::connection::SqliteConnection;
pub use self::query_builder::SqliteQueryBuilder;
