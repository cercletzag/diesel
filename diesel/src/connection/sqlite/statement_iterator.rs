use std::marker::PhantomData;

use backend::Sqlite;
use super::stmt::Statement;
use query_source::Queryable;
use types::{HasSqlType, FromSqlRow};

pub struct StatementIterator<ST, T> {
    stmt: Statement,
    _marker: PhantomData<(ST, T)>,
}

impl<ST, T> StatementIterator<ST, T> {
    pub fn new(stmt: Statement) -> Self {
        StatementIterator {
            stmt: stmt,
            _marker: PhantomData,
        }
    }
}

impl<ST, T> Iterator for StatementIterator<ST, T> where
    Sqlite: HasSqlType<ST>,
    T: Queryable<ST, Sqlite>,
{
    type Item = T;

    fn next(&mut self) -> Option<T> {
        self.stmt.step().map(|mut row| {
            let values = match T::Row::build_from_row(&mut row) {
                Ok(value) => value,
                Err(reason) => panic!("Error reading values {}", reason.description()),
            };
            T::build(values)
        })
    }
}
