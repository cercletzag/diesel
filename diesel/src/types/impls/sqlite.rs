use std::error::Error;

use backend::Sqlite;
use connection::sqlite::SqliteValue;
use super::option::UnexpectedNullError;
use types::{self, FromSql};

impl FromSql<types::VarChar, Sqlite> for String {
    fn from_sql(value: Option<&SqliteValue>) -> Result<Self, Box<Error>> {
        let text = try!(not_none!(value).read_text());
        Ok(text.into())
    }
}

impl FromSql<types::Binary, Sqlite> for Vec<u8> {
    fn from_sql(bytes: Option<&SqliteValue>) -> Result<Self, Box<Error>> {
        let bytes = not_none!(bytes).read_blob();
        Ok(bytes.into())
    }
}

impl FromSql<types::SmallInt, Sqlite> for i16 {
    fn from_sql(value: Option<&SqliteValue>) -> Result<Self, Box<Error>> {
        Ok(not_none!(value).read_integer() as i16)
    }
}

impl FromSql<types::Integer, Sqlite> for i32 {
    fn from_sql(value: Option<&SqliteValue>) -> Result<Self, Box<Error>> {
        Ok(not_none!(value).read_integer())
    }
}

impl FromSql<types::Bool, Sqlite> for bool {
    fn from_sql(value: Option<&SqliteValue>) -> Result<Self, Box<Error>> {
        Ok(not_none!(value).read_integer() != 0)
    }
}

impl FromSql<types::BigInt, Sqlite> for i64 {
    fn from_sql(value: Option<&SqliteValue>) -> Result<Self, Box<Error>> {
        Ok(not_none!(value).read_long())
    }
}

impl FromSql<types::Float, Sqlite> for f32 {
    fn from_sql(value: Option<&SqliteValue>) -> Result<Self, Box<Error>> {
        Ok(not_none!(value).read_double() as f32)
    }
}

impl FromSql<types::Double, Sqlite> for f64 {
    fn from_sql(value: Option<&SqliteValue>) -> Result<Self, Box<Error>> {
        Ok(not_none!(value).read_double())
    }
}
