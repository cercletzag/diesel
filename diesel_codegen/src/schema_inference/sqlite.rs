use diesel::*;
use diesel::expression::dsl::sql;
use diesel::sqlite::{SqliteConnection, Sqlite};
use diesel::types::{HasSqlType, FromSqlRow};
use syntax::ast;
use syntax::codemap::Span;
use syntax::ext::base::*;
use syntax::ptr::P;

use super::data_structures::*;

table!{
    pragma_table_info (cid){
        cid ->Integer,
        name -> VarChar,
        type_name -> VarChar,
        notnull -> Bool,
        dflt_value -> Nullable<VarChar>,
        pk -> Bool,
    }
}

pub fn get_table_data(conn: &SqliteConnection, table_name: &str)
    -> QueryResult<Vec<ColumnInformation>>
{
    let query = format!("PRAGMA TABLE_INFO('{}')", table_name);
    sql::<pragma_table_info::SqlType>(&query).load(conn)
}

fn is_text(type_name: &str) -> bool {
    type_name.contains("char")||
        type_name.contains("clob")||
        type_name.contains("text")
}

pub fn determine_column_type(cx: &mut ExtCtxt, attr: &ColumnInformation) -> P<ast::Ty> {
    let type_name=attr.type_name.to_lowercase();
    let tpe = if type_name.contains("int") {
        quote_ty!(cx, ::diesel::types::BigInteger)
    } else if is_text(&type_name) {
        quote_ty!(cx, ::diesel::types::Text)
    } else if type_name.contains("blob") || type_name.is_empty() {
        quote_ty!(cx, ::diesel::types::Binary)
    } else {
        quote_ty!(cx, ::diesel::types::Double)
    };

    if attr.nullable {
        quote_ty!(cx, Nullable<$tpe>)
    } else {
        tpe
    }
}

pub fn load_table_names(
    _cx: &mut ExtCtxt,
    _sp: Span,
    connection: &SqliteConnection,
) -> Result<Vec<String>, result::Error> {
    use diesel::prelude::*;
    use diesel::expression::dsl::sql;

    let query = select(sql::<types::VarChar>("name FROM sqlite_master"))
        .filter(sql::<types::Bool>("type='table' AND name NOT LIKE '\\_\\_%'"));
    query.load(connection)
}

struct FullTableInfo {
    _cid: i32,
    name: String,
    _type_name: String,
    _not_null: bool,
    _dflt_value: Option<String>,
    primary_key: bool,
}

impl<ST> Queryable<ST, Sqlite> for FullTableInfo where
    Sqlite: HasSqlType<ST>,
    (i32, String, String, bool, Option<String>, bool): FromSqlRow<ST, Sqlite>,
{
    type Row = (i32, String, String, bool, Option<String>, bool);

    fn build(row: Self::Row) -> Self {
        FullTableInfo {
            _cid: row.0,
            name: row.1,
            _type_name: row.2,
            _not_null: row.3,
            _dflt_value: row.4,
            primary_key: row.5,
        }
    }
}

pub fn get_primary_keys(conn: &SqliteConnection, table_name: &str) -> QueryResult<Vec<String>> {
    let query = format!("PRAGMA TABLE_INFO('{}')", table_name);
    let results = try!(sql::<pragma_table_info::SqlType>(&query)
        .load::<FullTableInfo>(conn));
    Ok(results.iter()
        .filter_map(|i| if i.primary_key { Some(i.name.clone()) } else { None })
        .collect())
}
