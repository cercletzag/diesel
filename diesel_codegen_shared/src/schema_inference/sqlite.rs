use diesel::*;
use diesel::expression::dsl::sql;
use diesel::sqlite::SqliteConnection;
use std::error::Error;

table! {
    sqlite_master (name) {
        name -> VarChar,
    }
}

pub fn load_table_names(connection: &SqliteConnection) -> Result<Vec<String>, Box<Error>> {
    use self::sqlite_master::dsl::*;

    sqlite_master.select(name)
        .filter(name.not_like("\\_\\_%").escape('\\'))
        .filter(name.not_like("sqlite%"))
        .filter(sql("type='table'"))
        .load(connection)
        .map_err(Into::into)
}
