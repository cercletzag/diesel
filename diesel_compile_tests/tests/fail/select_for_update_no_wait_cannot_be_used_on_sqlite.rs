extern crate diesel;

use diesel::*;
use diesel::sqlite::SqliteConnection;

table! {
    users {
        id -> Integer,
    }
}

fn main() {
    let conn = SqliteConnection::establish("").unwrap();
    users::table
        .for_update()
        .no_wait()
        .load(&conn)
        .unwrap();
}
