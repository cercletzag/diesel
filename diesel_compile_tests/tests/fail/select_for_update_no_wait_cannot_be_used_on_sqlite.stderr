error[E0277]: the trait bound `diesel::query_builder::locking_clause::ForUpdate: diesel::query_builder::QueryFragment<diesel::sqlite::Sqlite>` is not satisfied
  --> $DIR/select_for_update_no_wait_cannot_be_used_on_sqlite.rs:17:10
   |
17 |         .load(&conn)
   |          ^^^^ the trait `diesel::query_builder::QueryFragment<diesel::sqlite::Sqlite>` is not implemented for `diesel::query_builder::locking_clause::ForUpdate`
   |
   = help: the following implementations were found:
             <diesel::query_builder::locking_clause::ForUpdate as diesel::query_builder::QueryFragment<diesel::mysql::Mysql>>
             <diesel::query_builder::locking_clause::ForUpdate as diesel::query_builder::QueryFragment<diesel::pg::Pg>>
   = note: required because of the requirements on the impl of `diesel::query_builder::QueryFragment<diesel::sqlite::Sqlite>` for `diesel::query_builder::locking_clause::LockingClause<diesel::query_builder::locking_clause::ForUpdate, diesel::query_builder::locking_clause::NoWait>`
   = note: required because of the requirements on the impl of `diesel::query_builder::QueryFragment<diesel::sqlite::Sqlite>` for `diesel::query_builder::SelectStatement<users::table, diesel::query_builder::select_clause::DefaultSelectClause, diesel::query_builder::distinct_clause::NoDistinctClause, diesel::query_builder::where_clause::NoWhereClause, diesel::query_builder::order_clause::NoOrderClause, diesel::query_builder::LimitOffsetClause<diesel::query_builder::NoLimitClause, diesel::query_builder::NoOffsetClause>, diesel::query_builder::group_by_clause::NoGroupByClause, diesel::query_builder::locking_clause::LockingClause<diesel::query_builder::locking_clause::ForUpdate, diesel::query_builder::locking_clause::NoWait>>`
   = note: required because of the requirements on the impl of `diesel::query_dsl::LoadQuery<diesel::SqliteConnection, _>` for `diesel::query_builder::SelectStatement<users::table, diesel::query_builder::select_clause::DefaultSelectClause, diesel::query_builder::distinct_clause::NoDistinctClause, diesel::query_builder::where_clause::NoWhereClause, diesel::query_builder::order_clause::NoOrderClause, diesel::query_builder::LimitOffsetClause<diesel::query_builder::NoLimitClause, diesel::query_builder::NoOffsetClause>, diesel::query_builder::group_by_clause::NoGroupByClause, diesel::query_builder::locking_clause::LockingClause<diesel::query_builder::locking_clause::ForUpdate, diesel::query_builder::locking_clause::NoWait>>`

error[E0277]: the trait bound `diesel::query_builder::locking_clause::NoWait: diesel::query_builder::QueryFragment<diesel::sqlite::Sqlite>` is not satisfied
  --> $DIR/select_for_update_no_wait_cannot_be_used_on_sqlite.rs:17:10
   |
17 |         .load(&conn)
   |          ^^^^ the trait `diesel::query_builder::QueryFragment<diesel::sqlite::Sqlite>` is not implemented for `diesel::query_builder::locking_clause::NoWait`
   |
   = help: the following implementations were found:
             <diesel::query_builder::locking_clause::NoWait as diesel::query_builder::QueryFragment<diesel::mysql::Mysql>>
             <diesel::query_builder::locking_clause::NoWait as diesel::query_builder::QueryFragment<diesel::pg::Pg>>
   = note: required because of the requirements on the impl of `diesel::query_builder::QueryFragment<diesel::sqlite::Sqlite>` for `diesel::query_builder::locking_clause::LockingClause<diesel::query_builder::locking_clause::ForUpdate, diesel::query_builder::locking_clause::NoWait>`
   = note: required because of the requirements on the impl of `diesel::query_builder::QueryFragment<diesel::sqlite::Sqlite>` for `diesel::query_builder::SelectStatement<users::table, diesel::query_builder::select_clause::DefaultSelectClause, diesel::query_builder::distinct_clause::NoDistinctClause, diesel::query_builder::where_clause::NoWhereClause, diesel::query_builder::order_clause::NoOrderClause, diesel::query_builder::LimitOffsetClause<diesel::query_builder::NoLimitClause, diesel::query_builder::NoOffsetClause>, diesel::query_builder::group_by_clause::NoGroupByClause, diesel::query_builder::locking_clause::LockingClause<diesel::query_builder::locking_clause::ForUpdate, diesel::query_builder::locking_clause::NoWait>>`
   = note: required because of the requirements on the impl of `diesel::query_dsl::LoadQuery<diesel::SqliteConnection, _>` for `diesel::query_builder::SelectStatement<users::table, diesel::query_builder::select_clause::DefaultSelectClause, diesel::query_builder::distinct_clause::NoDistinctClause, diesel::query_builder::where_clause::NoWhereClause, diesel::query_builder::order_clause::NoOrderClause, diesel::query_builder::LimitOffsetClause<diesel::query_builder::NoLimitClause, diesel::query_builder::NoOffsetClause>, diesel::query_builder::group_by_clause::NoGroupByClause, diesel::query_builder::locking_clause::LockingClause<diesel::query_builder::locking_clause::ForUpdate, diesel::query_builder::locking_clause::NoWait>>`
